### DISCUSSION TITLE
Client generation: Scalar for BigDecimal

### DISCUSSION THREAD

**Discussion by huehnerlady** [ExpediaGroup/graphql-kotlin#1206] [Q&A]
HI,

after being able to generate a client with the help provided [here](https://github.com/ExpediaGroup/graphql-kotlin/discussions/1196) We now run into the issue, that we use java.math.BigDecimal in our GraphQL Schema, created with `com.expediagroup:graphql-kotlin-schema-generator`. We rely here on a central mode, which unfortunately is made out of Groovy classes (We currently have the main lanuage as groovy, but I try to introduce kotlin more and more), so changing BigDecimal is not possible.

In the Schema we can see the Scalar:

```graphwl
"Built-in java.math.BigDecimal"
scalar BigDecimal
```

But when using the graphql-kotlin client generator we end up with that type being just String. I could find a generated class:

```kotlin
/**
 * Built-in java.math.BigDecimal
 */
public typealias BigDecimal = String
```

Then I tried to use a customScalar:

```kotlin
    customScalars.add(GraphQLScalar("BigDecimal", "java.math.BigDecimal", "my.package.BigDecimalConverter"))

```

This seems to create a wrapper class though:

```kotlin
/**
 * Built-in java.math.BigDecimal
 */
public data class BigDecimal(
  public val value: java.math.BigDecimal
) {
  @JsonValue
  public fun rawValue() = converter.toJson(value)

  public companion object {
    public val converter: BigDecimalConverter = BigDecimalConverter()

    @JsonCreator
    @JvmStatic
    public fun create(rawValue: Any) = BigDecimal(converter.toScalar(rawValue))
  }
}
```

So again that is not the behaviour we want.

When I manually change the type alias to be `java.math.BigDecimal` it all works as expected. So can I somehow say in the config, that I just want a type alias? I was hoping the Comment would help there, as it is generated by graphql-kotlin, but that seems not to be enough?

Thanks for your help 🙂

**Comment by dariuszkuc** [comment]
Hello 👋 
Per [GraphQL spec](https://spec.graphql.org/June2018/#sec-Scalars) all `all GraphQL scalars are representable as strings` and while servers can use more appropriate primitives, clients have no knowledge what custom scalars represent. Since all scalars have to be representable as strings that is also the common default for all custom scalars. To keep logic simple and consistent, currently we don't support providing custom type aliases for custom scalars and always rely on the explicit converter to handle de-/serialization.

I'm thinking we could potentially support it by allowing to specify something like `GraphQLScalarTypeAlias` and default `GraphQLScalar` (with the converter).

**Reply by huehnerlady** [reply]
I understand that you basically mean instead of you generating `public typealias BigDecimal = String` I would do so and then configure this when generating the client?

That sounds like it would solve the issue nicely 👍🏻

**Reply by dariuszkuc** [reply]
No. I mean that could potentially be supported through config and the generator would generate appropriate type aliases

```
customScalars.add(
   GraphQLScalarTypeAlias("BigDecimal", "java.math.BigDecimal"),   // generates typealias
   GraphQLScalar("UUID", "java.util.UUID", "my.package.UUIDConverter") // generates wrapper using converter
)
```

**Reply by huehnerlady** [reply]
ah I see, that would be even better 👍🏻

**Reply by dariuszkuc** [reply]
@huehnerlady unfortunately I don't think we will be able to support it generically. See https://github.com/ExpediaGroup/graphql-kotlin/pull/1213#issuecomment-877414832 for details

If you have some ideas how to workaround those issues feel free to open a PR.

**Reply by huehnerlady** [reply]
I see. 

To go back to the problem might help figuring out a solution:

We have the problem that the Scalar Converters end up wrapping the Data in a class (this case BigDecimal with a value of java.math.BigDecimal. 


What we have basically is:

```kotlin

public data class Foo(
  public val property: BigDecimal? = null
)

/**
 * Built-in java.math.BigDecimal
 */
public data class BigDecimal(
  public val value: java.math.BigDecimal
) {
  @JsonValue
  public fun rawValue() = converter.toJson(value)

  public companion object {
    public val converter: BigDecimalConverter = BigDecimalConverter()

    @JsonCreator
    @JvmStatic
    public fun create(rawValue: Any) = BigDecimal(converter.toScalar(rawValue))
  }
}

```

So to instantiate an object you have to do: `Foo(property = BigDecimal(value = "10000".toBigDecimal()))`

What I would like to achieve is:
```kotlin

public data class Foo(
  public val property: java.math.BigDecimal? = null
)
```

So I then can instantiate an object with: `Foo(property = "10000".toBigDecimal())`



On the server side for we use BigDecimal [Coercing](https://github.com/graphql-java/graphql-java/blob/3139ccbb8a2313704518f39150223bda550d0500/src/main/java/graphql/schema/Coercing.java) to ensure the transformation of the Scalar into BigDecimal:

<details>
<summary>CustomSchemaGeneratorHooks for BigDecimal</summary>

We do it a bit complicated as we have Euro and Prozent in the code for better clarity.
```kotlin

class CustomSchemaGeneratorHooks : SchemaGeneratorHooks {

  override fun willGenerateGraphQLType(type: KType): GraphQLType? {
    val typeName = type.toString()

    return when {
      typeName.startsWith("my.package.Euro") -> graphqlEuroType
      typeName.startsWith("my.package.Prozent") -> graphqlProzentType
      else -> null
    }
  }
}

val graphqlEuroType: GraphQLScalarType = newScalar()
  .name("Euro")
  .description("BigDecimal wrapped as Euro for better clarity")
  .coercing(GraphqlBigDecimalCoercing())
    .build()

val graphqlProzentType: GraphQLScalarType = newScalar()
  .name("Prozent")
  .description("BigDecimal wrapped as Prozent for better clarity")
  .coercing(GraphqlBigDecimalCoercing())
    .build()
```
</details>

So maybe it is somehow else possible to ensure that the value itself is not wrapped into a class but the type straight away, similar with coercing on the server side?

Going back to [your comment](https://github.com/ExpediaGroup/graphql-kotlin/pull/1213#issuecomment-877414832) maybe a converter is fine, as long as the type used in the generated data class is not wrapped?

**Reply by dariuszkuc** [reply]
I was playing around with annotating just the fields and I think https://github.com/ExpediaGroup/graphql-kotlin/pull/1215 could work (e.g. see https://github.com/dariuszkuc/graphql-kotlin/blob/scalars/plugins/client/graphql-kotlin-client-generator/src/test/data/jackson/custom_scalars/customscalarquery/ScalarWrapper.kt#L18).

**Reply by huehnerlady** [reply]
That sounds very promising, many thanks for looking into this for us ❤️

**Reply by dariuszkuc** [reply]
@huehnerlady we released the change in `5.0.0-alpha.2`

**Reply by huehnerlady** [reply]
@dariuszkuc great, thank you so much for your help. I have to say I am very glad that we use your library as I feel the support and the responsiveness is absolutely amazing ❤️ 

As we use that in production unfortunately all I can do with an alpha version is test it, but not use it. Do you have an estimate when you plan to release version 5?

**Reply by smyrick** [reply]
@dariuszkuc I don't have any major features left that *need* to get in. GraphQL Java is getting ready to release v17 so we could wait for that https://github.com/graphql-java/graphql-java/releases

**Reply by huehnerlady** [reply]
@dariuszkuc @smyrick any update on when you plan to release 5.0? :)
